p3team-05/man/man8/fbdctl.8
.TH FBDCTL 8
.SH NAME
fbdctl \- Faulty Block Device rule management interface
.SH SYNOPSIS
\fBfbdctl\fR \fBadd\fR [\fB-d\fR \fIdevice\fR]
[\fB-a\fR \fIstart\fR[\fB-\fR\fIend\fR]] [\fB-s\fR \fIskip\fR]
[\fB-c\fR \fIcount\fR] [\fB-rw\fR] \fIaction\fR [\fIparams\fR]
.PP
\fBfbdctl\fR \fBdel\fR [\fB-d\fR \fIdevice\fR] \fIrulenum\fR
.PP
\fBfbdctl\fR \fBlist\fR [\fB-d\fR \fIdevice\fR]
.SH DESCRIPTION
The Faulty Block Device (FBD) driver is an interposing block device driver
which can simulate certain disk-level I/O corruption and errors, based on a
user-provided set of rules. The \fBfbdctl\fR tool allows one to add, delete,
and list rules on a running FBD driver instance.
.PP
The \fBadd\fR subcommand adds a new rule, which will perform a predefined
faulty action on a disk transfer when triggered. See the ACTIONS subsection
below. The \fBdel\fR subcommands deletes an existing rule, based on its rule
number. All currently active rules and their corresponding rule numbers can be
viewed with the \fBlist\fR subcommand.
.SH OPTIONS
.TP 10
\fB-d\fR \fIdevice\fR
By default, \fBfbdctl\fR operates on \fB/dev/fbd\fR. With this option, one can
specify a different device node. This is useful when using multiple FBD
instances at the same time. The user would have to create extra device nodes
first in that case.
.TP 10
\fB-a\fR [\fIstart\fR[\fB-\fR\fIend\fR]]
When adding a rule, this option specifies the disk address range for which the
rule triggers. That is, the rule will trigger when an I/O operation overlaps
with the given range. Both \fIstart\fR and \fIend\fR are expected to be
hexadecimal numbers, without a "0x" prefix. The \fIend\fR address is exclusive.
If no \fIend\fR address is given, the rule will affect the disk from the
starting address to the disk end. If this option is not provided at all,
the rule will affect the entire disk.
.TP 10
\fB-s\fR \fIskip\fR
This option makes the new rule refrain from triggering for the given number
of times it matches. The \fIskip\fR value must be a positive decimal number.
If this option is omitted, the value is set to zero, meaning the rule will
start triggering immediately.
.TP 10
\fB-c\fR \fIcount\fR
This option makes the new rule trigger for this many I/O operations when
matched, after which it will be removed automatically. The \fIcount\fR value
must be a positive decimal number. If this option is omitted, or a value of
zero is given, the rule is permanent and will not be removed automatically.
.TP 10
\fB-r\fR, \fB-w\fR
These options allow one to make the new rule trigger on read or write
operations only. By default, or when both are specified, the new rule will
trigger for both read and write I/O operations.
.SH ACTIONS
The following actions are supported. They are executed when the rule matches
for a transfer request, and is triggered as a result. Note that the exact
meaning of the rule's disk address range (as given by the \fB-a\fR option)
depends on the action type.
.TP 10
\fBcorrupt\fR [\fBzero\fR|\fBpersist\fR|\fBrandom\fR]
In the part of the transfer that matches the disk address range given for the
rule (i.e., the intersection of the rule range and the transfer range), the
data will be corrupted. The following corruption policies are supported: the
data is set to \fBzero\fR, it is \fBpersist\fRently set to the same garbage
data for the same disk locations, or it is set to different \fBrandom\fR data
in every transfer.
.TP 10
\fBerror\fR [\fBOK\fR|\fBEIO\fR]
Only the part of the transfer up to the start of the rule's disk address range
will be performed, after which the given error code is returned. The \fBOK\fR
code effectively simulates an end-of-disk condition, whereas the \fBEIO\fR code
simulates generic disk-level I/O failure.
.TP 10
\fBmisdir\fR \fIstart\fR\fB-\fR\fIend\fR \fIalign\fR
Transfer requests that match this rule, will be \fBmisdirected\fR in their
entirety, to a random location in the given address range, and with the given
disk byte alignment within that range. The \fIstart\fR and \fIend\fR parameters
are specified just like the \fB-a\fR option, although the \fIend\fR parameter
is compulsory here (since the driver does not know the disk end). The
\fIalign\fR value must be a positive nonzero decimal number, and should be a
multiple of the medium sector size; a typical value would be 4096.
.TP 10
\fBlost\fR
Transfer requests that match this rule, will be \fBlost\fR in their entirety.
That is, they will not actually be performed, and yet report successful
completion.
.TP 10
\fBtorn\fR \fIlead\fR
Transfer requests that match this rule, will be \fBtorn\fR: only the first
\fIlead\fR bytes of the transfer request will actually be performed, and yet
completion of the full transfer is reported. The \fIlead\fR value must be a
positive nonzero decimal number. A \fBtorn\fR action with a \fIlead\fR value of
zero would be the same as the \fBlost\fR action.
.SH EXAMPLES
.TP 10
.B fbdctl add -a 2000-3000 corrupt zero
# Zero out the 4096 bytes starting from disk address 0x2000 in any transfer
that involves any of those bytes.
.TP 10
.B fbdctl add -s 9 -c 1 -r error EIO
# Fail the tenth read request with an I/O error.
.TP 10
.B fbdctl add -a A0000-B0000 -w misdir D0000-E0000 4096
# Misdirect write requests that overlap with the first range, to fall somewhere
in the second range, at 4K-block-granular alignment.
.SH AUTHOR
David van Moolenbroek <david@minix3.org>


p3team-05/commands/fbdctl/fbdctl.c
/* fbdctl - FBD control tool - by D.C. van Moolenbroek */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <minix/ioctl.h>
#include <minix/u64.h>
#include <sys/ioc_fbd.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

static int usage(char *name)
{
	printf("usage:\n");
	printf("  %s list\n", name);
	printf("  %s add [-a start[-end]] [-s skip] [-c count] [-rw] "
		"<action> [params]\n", name);
	printf("  %s del N\n", name);
	printf("\n");
	printf("actions and params:\n");
	printf("  corrupt [zero|persist|random]\n");
	printf("  error [OK|EIO]\n");
	printf("  misdir <start>-<end> <align>\n");
	printf("  lost\n");
	printf("  torn <lead>\n");
	printf("use %s -d <device> to specify a device other than /dev/fbd\n",
		name);

	return EXIT_FAILURE;
}

static void print_rule(struct fbd_rule *rule)
{
	printf("%-2d %04lX%08lX-%04lX%08lX %-4d %-5d %c%c ",
		rule->num, ex64hi(rule->start), ex64lo(rule->start),
		ex64hi(rule->end), ex64lo(rule->end), rule->skip,
		rule->count, (rule->flags & FBD_FLAG_READ) ? 'r' : ' ',
		(rule->flags & FBD_FLAG_WRITE) ? 'w' : ' ');

	switch (rule->action) {
	case FBD_ACTION_CORRUPT:
		printf("%-7s ", "corrupt");
		switch (rule->params.corrupt.type) {
		case FBD_CORRUPT_ZERO: printf("zero"); break;
		case FBD_CORRUPT_PERSIST: printf("persist"); break;
		case FBD_CORRUPT_RANDOM: printf("random"); break;
		default: printf("<unknown>");
		}
		break;

	case FBD_ACTION_ERROR:
		printf("%-7s ", "error");

		switch (rule->params.error.code) {
		case 0:
			printf("OK");
			break;
		case EIO:
		case -EIO:
			printf("EIO");
			break;
		default:
			printf("%d", rule->params.error.code);
		}

		break;

	case FBD_ACTION_MISDIR:
		printf("%-7s %04lX%08lX-%04lX%08lX %u",
			"misdir", ex64hi(rule->params.misdir.start),
			ex64lo(rule->params.misdir.start),
			ex64hi(rule->params.misdir.end),
			ex64lo(rule->params.misdir.end),
			rule->params.misdir.align);
		break;

	case FBD_ACTION_LOSTTORN:
		if (rule->params.losttorn.lead > 0)
			printf("%-7s %u", "torn",
				rule->params.losttorn.lead);
		else
			printf("%-7s", "lost");
	}

	printf("\n");
}

static int do_list(int fd)
{
	struct fbd_rule rule;
	int i;

	printf("N  Start        End          Skip Count RW Action  Params\n");

	for (i = 1; ; i++) {
		rule.num = i;

		if (ioctl(fd, FBDCGETRULE, &rule) < 0) {
			if (errno == ENOENT)
				continue;
			break;
		}

		print_rule(&rule);
	}

	return EXIT_SUCCESS;
}

static int scan_hex64(char *input, u64_t *val)
{
	u32_t lo, hi;
	char buf[9];
	int len;

	len = strlen(input);

	if (len < 1 || len > 16) return 0;

	if (len > 8) {
		memcpy(buf, input, len - 8);
		buf[len - 8] = 0;
		input += len - 8;

		hi = strtoul(buf, NULL, 16);
	}
	else hi = 0;

	lo = strtoul(input, NULL, 16);

	*val = make64(lo, hi);

	return 1;
}

static int scan_range(char *input, u64_t *start, u64_t *end, int need_end)
{
	char *p;

	if ((p = strchr(input, '-')) != NULL) {
		*p++ = 0;

		if (!scan_hex64(p, end)) return 0;
	}
	else if (need_end) return 0;

	return scan_hex64(input, start);
}

static int do_add(int fd, int argc, char **argv, int off)
{
	struct fbd_rule rule;
	int c, r;

	memset(&rule, 0, sizeof(rule));

	while ((c = getopt(argc-off, argv+off, "a:s:c:rw")) != EOF) {
		switch (c) {
		case 'a':
			if (!scan_range(optarg, &rule.start, &rule.end, 0))
				return usage(argv[0]);
			break;
		case 's':
			rule.skip = atoi(optarg);
			break;
		case 'c':
			rule.count = atoi(optarg);
			break;
		case 'r':
			rule.flags |= FBD_FLAG_READ;
			break;
		case 'w':
			rule.flags |= FBD_FLAG_WRITE;
			break;
		default:
			return usage(argv[0]);
		}
	}

	optind += off; /* compensate for the shifted argc/argv */

	if (optind >= argc) return usage(argv[0]);

	/* default to reads and writes */
	if (!rule.flags) rule.flags = FBD_FLAG_READ | FBD_FLAG_WRITE;

	if (!strcmp(argv[optind], "corrupt")) {
		if (optind+1 >= argc) return usage(argv[0]);

		rule.action = FBD_ACTION_CORRUPT;

		if (!strcmp(argv[optind+1], "zero"))
			rule.params.corrupt.type = FBD_CORRUPT_ZERO;
		else if (!strcmp(argv[optind+1], "persist"))
			rule.params.corrupt.type = FBD_CORRUPT_PERSIST;
		else if (!strcmp(argv[optind+1], "random"))
			rule.params.corrupt.type = FBD_CORRUPT_RANDOM;
		else return usage(argv[0]);
	}
	else if (!strcmp(argv[optind], "error")) {
		if (optind+1 >= argc) return usage(argv[0]);

		rule.action = FBD_ACTION_ERROR;

		if (!strcmp(argv[optind+1], "OK"))
			rule.params.error.code = 0;
		else if (!strcmp(argv[optind+1], "EIO")) {
			if (EIO > 0)
				rule.params.error.code = -EIO;
			else
				rule.params.error.code = EIO;
		}
		else return usage(argv[0]);
	}
	else if (!strcmp(argv[optind], "misdir")) {
		if (optind+2 >= argc) return usage(argv[0]);

		rule.action = FBD_ACTION_MISDIR;

		if (!scan_range(argv[optind+1], &rule.params.misdir.start,
				&rule.params.misdir.end, 1))
			return usage(argv[0]);

		rule.params.misdir.align = atoi(argv[optind+2]);

		if ((int) rule.params.misdir.align <= 0)
			return usage(argv[0]);
	}
	else if (!strcmp(argv[optind], "lost")) {
		rule.action = FBD_ACTION_LOSTTORN;

		rule.params.losttorn.lead = 0;
	}
	else if (!strcmp(argv[optind], "torn")) {
		if (optind+1 >= argc) return usage(argv[0]);

		rule.action = FBD_ACTION_LOSTTORN;

		rule.params.losttorn.lead = atoi(argv[optind+1]);

		if ((int) rule.params.losttorn.lead <= 0)
			return usage(argv[0]);
	}
	else return usage(argv[0]);

#if DEBUG
	print_rule(&rule);
#endif

	r = ioctl(fd, FBDCADDRULE, &rule);

	if (r < 0) {
		perror("ioctl");

		return EXIT_FAILURE;
	}

	printf("Added rule %d\n", r);

	return EXIT_SUCCESS;
}

static int do_del(int fd, int argc, char **argv, int off)
{
	fbd_rulenum_t num;

	if (argc < off + 2)
		return usage(argv[0]);

	num = atoi(argv[off + 1]);

	if (ioctl(fd, FBDCDELRULE, &num)) {
		perror("ioctl");

		return EXIT_FAILURE;
	}

	printf("Deleted rule %d\n", num);

	return EXIT_SUCCESS;
}

int main(int argc, char **argv)
{
	int r, fd, off = 1;
	char *dev = "/dev/fbd";

	if (argc < 2)
		return usage(argv[0]);

	if (!strcmp(argv[1], "-d")) {
		if (argc < 4)
			return usage(argv[0]);

		dev = argv[2];

		off += 2;
	}

	fd = open(dev, O_RDONLY);
	if (fd < 0) {
		perror(dev);

		return EXIT_FAILURE;
	}

	else if (!strcmp(argv[off], "list"))
		r = do_list(fd);
	else if (!strcmp(argv[off], "add"))
		r = do_add(fd, argc, argv, off);
	else if (!strcmp(argv[off], "del"))
		r = do_del(fd, argc, argv, off);
	else
		r = usage(argv[0]);

	close(fd);

	return r;
}
